src/main/java/uk/gov/hmcts/reform/idam/web/strategic/StepUpAuthenticationZuulFilter.java

```diff
@@ -0,0 +1,136 @@
+        this.idamSessionCookieName = configurationProperties.getStrategic().getSession().getIdamSessionCookie();
+        this.spiService = spiService;
+    }
+
+    @Override
+    public String filterType() {
+        return PRE_TYPE;
+    }
+
+    @Override
+    public int filterOrder() {
+        return 1;
+    }
+
+    @Override
+    public boolean shouldFilter() {
+        final HttpServletRequest request = RequestContext.getCurrentContext().getRequest();
+        return isAuthorizeRequest(request) && hasSessionCookie(request);
+    }
+
+
+    @Override
+    public Object run() throws ZuulException {
+        final RequestContext ctx = RequestContext.getCurrentContext();
+        final HttpServletRequest request = ctx.getRequest();
+
+        log.info("StepUp filter triggered.");
+
+        final String tokenId = getSessionToken(request);
+
+        final URI loginRedirectUri;
+        try {
+            final URIBuilder redirectUriBuilder = new URIBuilder().setPath("/login");
+            copyRequestParameters(redirectUriBuilder, request);
+            loginRedirectUri = redirectUriBuilder.build();
+        } catch (final URISyntaxException e) {
+            throw zuulError("Cannot generate loginRedirectUri");
+        }
+
+        try {
+            final String originIp = ObjectUtils.defaultIfNull(request.getHeader(X_FORWARDED_FOR), request.getRemoteAddr());
+            final String redirectUri = request.getParameter(MvcKeys.REDIRECT_URI);
+            final ApiAuthResult authenticationResult = spiService.authenticate(tokenId, redirectUri, originIp);
+            if (!authenticationResult.isSuccess()) {
+                throw zuulError("Authentication failed: " + authenticationResult.getErrorCode().name());
+            }
+            // redirect to the login page for re-authentication
+            if (authenticationResult.requiresMfa()) {
+                ctx.setResponseStatusCode(HttpServletResponse.SC_MOVED_TEMPORARILY);
+                ctx.getResponse().setHeader(HttpHeaders.LOCATION, loginRedirectUri.toString());
+                ctx.setSendZuulResponse(false);
+                return null;
+            }
+            // continue as usual
+            ctx.setSendZuulResponse(true);
+            return null;
+        } catch (final JsonProcessingException e) {
+            log.error(ZUUL_PROCESSING_ERROR, e);
+            throw zuulError(ZUUL_PROCESSING_ERROR);
+        }
+    }
+
+    protected String getSessionToken(@Nonnull final HttpServletRequest request) {
+        return Arrays.stream(getCookiesFromRequest(request))
+            .filter(cookie -> idamSessionCookieName.equals(cookie.getName()))
+            .map(Cookie::getValue)
+            .findAny()
+            .orElseThrow();
+    }
+
+    protected boolean isAuthorizeRequest(@Nonnull final HttpServletRequest request) {
+        return request.getRequestURI().contains("/o/authorize") && "post".equalsIgnoreCase(request.getMethod());
```

Comment:
Normally, the Authorization request is `GET /o/authorize`. Are we sure we need a `post` here?

Reply:
Fixed